### DISCUSSION TITLE
Integration testing of the generated endpoint only

### DISCUSSION THREAD

**Discussion by keetron** [ExpediaGroup/graphql-kotlin#1256] [Q&A] [answered]
When using `@SpringBootTest`it allows the user to start up the spring context with a limited amount of classes which can be useful for isolated integration testing the inbound and outbound parts of the service. 
An example how this would look like: 
```
@SpringBootTest(classes = [ExposedGraphQL::class], webEnvironment = RANDOM_PORT)
@AutoConfigureWebTestClient
@EnableAutoConfiguration
class ExposedGraphQLIT(@Autowired private val testClient: WebTestClient)  {
((tests go here))
}
```

This however is throwing errors:
```
nested exception is com.expediagroup.graphql.generator.exceptions.TypeNotSupportedException: Cannot convert java.time.Instant since it is not a valid GraphQL type or outside the supported packages "[your.package]"
``` 

Adding the java.time package to the `graphql.packages` list is not possible as then the following error is thrown:
```
nested exception is com.expediagroup.graphql.generator.exceptions.InvalidInputFieldTypeException: Argument cannot be an interface or a union, parameter #1 arg0 of fun java.time.temporal.TemporalAccessor.range(java.time.temporal.TemporalField!): java.time.temporal.ValueRange!
```

Of course, I could start up the entire application, spin up docker containers for database and redis, start up wiremock for downstream services and basically end up with a slow and complicated test and I do not want to do that to test only the _configuration_ of the graphql controller that is generated by the graphql-kotlin plugin.

How to enable this test, what am I doing wrong here?

**Comment by smyrick** [comment, marked as answer]
It looks like you are trying to return `TemporalAccessor` somewhere in the schema so it is taking all the functions in this class and running them through the generator. The method `TemporalAccessor.range` has one argument and that happens to be an interface. In Kotlin this is valid, but in GraphQL you can not use interfaces as input objects. So the resolution here is to either use a type different than `TemporalAccessor` or modify the behaviour with hooks to ignore this method.

I would suggest not using any of the base java classes in your schema unless you are doing something with custom scalars. Avoiding these types means your API schema is better controlled and more focuses on returning the exact data the client needs rather than having the ability to access every method on `TemporalAccessor`

**Reply by keetron** [reply]
As it turns out the application indeed has customised SchemaGenerationHooks, the challenge now is how to wire those into the test.

**Reply by keetron** [reply]
Update a few hours later:
With the trusted `@TestConfiguration` a configuration was set up to be able to load the bean for the custom hooks. This allows me to run controller tests without having to spin up the entire context.

**Comment by Keerthan22** [comment]
Hello @keetron I'm facing a similar issue as this but in a different setup using Quarkus and exception with `com.expediagroup.graphql.generator.scalars.ID` type, can you elaborate on what exactly you configured to get the integration tests working?

**Reply by keetron** [reply]
No sorry, this is rather long ago, I left that role and never encountered a similar problem since.

**Reply by smyrick** [reply]
@Keerthan22 Check to make sure that you are using your custom hooks in the integration tests

**Reply by Keerthan22** [reply]
@smyrick Currently I'm spinning up the resource using `@QuarkusTest` & testing the same using `RestAssured`, although I expect it to work OOTB, I'm not sure whats wrong.
On a side note: The application works as expected when built as a Quarkus application, its only the Integration tests with `@QuarkusTest` annotation that have this issue

**Reply by smyrick** [reply]
I am not familiar with Quarkus, you can look at the examples folder for Spring and how that imports the things for `@SpringTest`

https://github.com/ExpediaGroup/graphql-kotlin/blob/ad93137f4906e67a45f403039876a681b1c9e2a0/examples/server/spring-server/src/main/kotlin/com/expediagroup/graphql/examples/server/spring/Application.kt#L42

If you are getting a `com.expediagroup.graphql.generator.exceptions.TypeNotSupportedException` in test and not in app then to me that idicates that you hooks are not being properly configured for the test
